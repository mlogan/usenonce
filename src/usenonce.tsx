
import React, { ReactNode, createContext, useContext, useState } from 'react'

import Alea from 'alea'

type AleaType = ReturnType<typeof Alea>

export class NonceGenerator {
  _rng: AleaType
  constructor (seed: string) {
    // @ts-ignore
    this._rng = new Alea(seed)
  }

  next (): string {
    return this._rng().toString(36).substring(2, 15)
  }
}

export const NonceGeneratorCtx = createContext<NonceGenerator | null>(null)

export const NonceProvider: React.FC<{
  children: ReactNode,
  seed?: string
}> = ({children, seed = 'default seed'}) => {

  const generator = new NonceGenerator(seed)
  return <NonceGeneratorCtx.Provider value={generator}>
    {children}
  </NonceGeneratorCtx.Provider>
}

export const useNonce = (prefix: string = '') => {

  const nonceGenerator = useContext(NonceGeneratorCtx)

  if (nonceGenerator == null) {
    throw new Error("useNonce() must be called from within <NonceProvider>")
  }

  const [nonce] = useState(() => {
    return nonceGenerator.next()
  })

  return prefix + nonce
}
