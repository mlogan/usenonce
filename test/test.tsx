import React from 'react'
import { ListOfNonces } from './test-components'
import { NonceProvider } from '../src/usenonce'

import Adapter from 'enzyme-adapter-react-16'
import { configure, mount } from 'enzyme'
import toJSON from 'enzyme-to-json'

configure({ adapter: new Adapter() })

test('deterministic rendering', () => {
  const wrapper = mount(
    <NonceProvider>
      <ListOfNonces/>
    </NonceProvider>
  )

  expect(toJSON(wrapper)).toMatchSnapshot()

  wrapper.find('button').last().simulate('click')
  wrapper.update()

  // Make sure that nonces are stable across re-renders
  expect(toJSON(wrapper)).toMatchSnapshot()
})
