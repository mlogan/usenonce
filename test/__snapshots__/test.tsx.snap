// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deterministic rendering 1`] = `
<NonceProvider>
  <ListOfNonces>
    <button
      onClick={[Function]}
    />
    <NonceComponent
      key="0"
    >
      <div>
        xtufb06tirp
      </div>
    </NonceComponent>
    <NonceComponent
      key="1"
    >
      <div>
        0krkr4iueu2
      </div>
    </NonceComponent>
    <NonceComponent
      key="2"
    >
      <div>
        4q1om0duunl
      </div>
    </NonceComponent>
  </ListOfNonces>
</NonceProvider>
`;

exports[`deterministic rendering 2`] = `
<NonceProvider>
  <ListOfNonces>
    <button
      onClick={[Function]}
    />
    <NonceComponent
      key="0"
    >
      <div>
        xtufb06tirp
      </div>
    </NonceComponent>
    <NonceComponent
      key="1"
    >
      <div>
        0krkr4iueu2
      </div>
    </NonceComponent>
    <NonceComponent
      key="2"
    >
      <div>
        4q1om0duunl
      </div>
    </NonceComponent>
    <NonceComponent
      key="3"
    >
      <div>
        e6924wtecvu
      </div>
    </NonceComponent>
  </ListOfNonces>
</NonceProvider>
`;
